pyoneer/layers/dropout_impl.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+
+
+class ConcreteDropout(tf.keras.layers.Wrapper):
+    """
+    Implementation of concrete dropout (Gal et al. 2017).
+
+    Args:
+        layer: Apply dropout to the inputs of `layer` when called.
+        initial_rate: Initial dropout rate to apply.
+
+    Usage:
+
+        base_layer = tf.keras.layers.Dense(units=3)
+        dropout_layer = ConcreteDropout(base_layer)
+        dropout_layer(inputs)
+
+    References:
+        Concrete Dropout. Gal et al. 2017. https://arxiv.org/abs/1705.07832
+    """
+
+    def __init__(
+        self,
+        layer,
+        initial_rate=0.1,
+        kernel_regularization=1e-6,
+        dropout_regularization=1e-5,
+        noise_shape=None,
+        seed=None,
+        **kwargs
+    ):
+        super(ConcreteDropout, self).__init__(layer, **kwargs)
+        self.initial_rate = initial_rate
+        self.kernel_regularization = kernel_regularization
+        self.dropout_regularization = dropout_regularization
+        self.noise_shape = noise_shape
+        self.seed = seed
+        self.supports_masking = True
+        self.rate_logit = None
+
+    def build(self, input_shape):
+        super(ConcreteDropout, self).build(input_shape)
+
+        initial_rate_norm = np.log(self.initial_rate) - np.log(1.0 - self.initial_rate)
+        rate_initializer = tf.keras.initializers.Constant(initial_rate_norm)
+
+        self.rate_logit = self.add_weight(
+            name="rate_logit", shape=(), initializer=rate_initializer, trainable=True
+        )
+
+        if self.noise_shape is None:
+            self.noise_shape = input_shape
+
+        input_dim = np.prod(self.noise_shape[1:])
+
+        kernel_regularizer = (
```

Comment:
Can this be implemented as a [tf.keras.reguralizers.Regularizer](https://www.tensorflow.org/api_docs/python/tf/keras/regularizers/Regularizer) and the model should use this (you could add it in the module here)? It might make this semantically easier to understand

Reply:
Maybe it boils down to having a `ConcreteDropoutRegularizer` and a `ConcreteKernelRegularizer` or something. Ultimately the losses get summed together.

Reply:
Yeah, I’ll look into that. The challenge is both regularizers need access to the current dropout rate and the downstream kernel. Any thoughts on API?

```
droput_layer = ConcreteDropout()
droput_regularizer = ConcreteDropoutRegularizer(dropout_layer.rate_logit)  # NOTE: rate_logit doesn’t exist until `build` has been called
kernel_regularizer = ConcretKernelRegularizer(dropout_layer.rate_logit)
```

Reply:
Hmm, maybe this is something that should be added later. I'm not sure about the implementation details from the paper, but that may help me understand it better.